### ISSUE TITLE
ICE compiling async fn with multiple early returns, but missing final return

### ISSUE THREAD

**Issue by nate-onesignal** [rust-lang/rust#66533]
I triggered an ICE when using an async fn that was missing its final return, but had previous early returns in branches. I'm aware that the code should not compile successfully, but it seems that it shouldn't trigger an ICE either.

I tried this code:

```
async fn bar() -> i32 {
    if true {
        return 1;
    }
    
    if false {
        return 2;
    }
}
```

I expected to see this happen: 

Normal compiler error telling me that the function is missing a return value

Instead, this happened: 

```
error: internal compiler error: src/librustc/ty/context.rs:554: node_type: no type for node `expr { } (hir_id=HirId { owner: DefIndex(12), local_id: 19 })`

thread 'rustc' panicked at 'Box<Any>', src/librustc_errors/lib.rs:812:9
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace.

note: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust/blob/master/CONTRIBUTING.md#bug-reports

note: rustc 1.39.0 (4560ea788 2019-11-04) running on x86_64-unknown-linux-gnu

note: compiler flags: -C codegen-units=1 -C debuginfo=2 --crate-type lib

note: some of the compiler flags provided by cargo are hidden
```

## Meta

`rustc --version --verbose`: `rustc 1.39.0 (4560ea788 2019-11-04)`

Backtrace:

```
λ RUSTC_LOG=ty::context=debug RUST_BACKTRACE=full rustc foo.rs          
error[E0670]: `async fn` is not permitted in the 2015 edition
 --> foo.rs:1:1
  |
1 | async fn bar() -> i32 {
  | ^^^^^

error[E0601]: `main` function not found in crate `foo`
 --> foo.rs:1:1
  |
1 | / async fn bar() -> i32 {
2 | |     if true {
3 | |         return 1;
4 | |     }
... |
8 | |     }
9 | | }
  | |_^ consider adding a `main` function to `foo.rs`

error[E0317]: if may be missing an else clause
 --> foo.rs:6:5
  |
6 | /     if false {
7 | |         return 2;
8 | |     }
  | |_____^ expected (), found integer
  |
  = note: expected type `()`
             found type `{integer}`
  = note: `if` expressions without `else` evaluate to `()`
  = help: consider adding an `else` block that evaluates to the expected type

error: internal compiler error: src/librustc/ty/context.rs:554: node_type: no type for node `expr { } (hir_id=HirId { owner: DefIndex(12), local_id: 19 })`

thread 'rustc' panicked at 'Box<Any>', src/librustc_errors/lib.rs:812:9
stack backtrace:
   0:        0x103a2d165 - <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt::h66f38d4abb2e41c1
   1:        0x103a63e10 - core::fmt::write::h4f487e714088986d
   2:        0x103a207bb - std::io::Write::write_fmt::h7bd7b6fe8b47c5fb
   3:        0x103a314aa - std::panicking::default_hook::{{closure}}::heee79636c241547c
   4:        0x103a311b5 - std::panicking::default_hook::hfcbd07059d15441e
   5:        0x100fc4762 - rustc_driver::report_ice::h5ad941de17a16320
   6:        0x103a31ce2 - std::panicking::rust_panic_with_hook::h0c4b67125f55410a
   7:        0x102afce51 - std::panicking::begin_panic::h15f5ca0459a0afbf
   8:        0x102b11d37 - rustc_errors::HandlerInner::bug::h3ae5a8494001dbc6
   9:        0x102b10bdd - rustc_errors::Handler::bug::h926b6bc1e5e81222
  10:        0x102424dfb - rustc::util::bug::opt_span_bug_fmt::{{closure}}::hc9e1ba37629e505a
  11:        0x10241f306 - rustc::ty::context::tls::with_opt::{{closure}}::h0f2d2c67a51c6ea3
  12:        0x10241f294 - rustc::ty::context::tls::with_context_opt::hffb989149536908e
  13:        0x10241f2c2 - rustc::ty::context::tls::with_opt::h8dbbf9943655c843
  14:        0x102424d18 - rustc::util::bug::opt_span_bug_fmt::haea6cb90fd22dc2a
  15:        0x102424c6b - rustc::util::bug::bug_fmt::hd6a2040ffa44f924
  16:        0x102616caf - rustc::ty::context::TypeckTables::node_type::{{closure}}::h7013b08e9165475c
  17:        0x102617263 - rustc::ty::context::TypeckTables::expr_ty::h97963f0b78948fbd
  18:        0x10165102e - <rustc_typeck::check::generator_interior::InteriorVisitor as rustc::hir::intravisit::Visitor>::visit_expr::h3b58dc6fbd3e7928
  19:        0x10166ccbf - rustc::hir::intravisit::walk_expr::h05f1735da88485b2
  20:        0x101650f7c - <rustc_typeck::check::generator_interior::InteriorVisitor as rustc::hir::intravisit::Visitor>::visit_expr::h3b58dc6fbd3e7928
  21:        0x101650f7c - <rustc_typeck::check::generator_interior::InteriorVisitor as rustc::hir::intravisit::Visitor>::visit_expr::h3b58dc6fbd3e7928
  22:        0x101650f7c - <rustc_typeck::check::generator_interior::InteriorVisitor as rustc::hir::intravisit::Visitor>::visit_expr::h3b58dc6fbd3e7928
  23:        0x101650f7c - <rustc_typeck::check::generator_interior::InteriorVisitor as rustc::hir::intravisit::Visitor>::visit_expr::h3b58dc6fbd3e7928
  24:        0x10165061d - rustc_typeck::check::generator_interior::resolve_interior::h94e9b683be16f5d7
  25:        0x1014ce8b1 - rustc_typeck::check::FnCtxt::resolve_generator_interiors::h3f88be1dbc9c6168
  26:        0x10161c575 - rustc::ty::context::GlobalCtxt::enter_local::h5020ef7c3f2707ac
  27:        0x1014c0a5c - rustc_typeck::check::typeck_tables_of::h1d829091428d57d9
  28:        0x10150dcda - rustc::ty::query::__query_compute::typeck_tables_of::h45dc00b6674693bf
  29:        0x10159859b - rustc::ty::query::<impl rustc::ty::query::config::QueryAccessors for rustc::ty::query::queries::typeck_tables_of>::compute::had7daa30ac2f087b
  30:        0x1015381db - rustc::dep_graph::graph::DepGraph::with_task_impl::h422fc241c771cd19
  31:        0x1015e5f16 - rustc::ty::query::plumbing::<impl rustc::ty::context::TyCtxt>::get_query::ha62632bdf7526a3d
  32:        0x10150dcda - rustc::ty::query::__query_compute::typeck_tables_of::h45dc00b6674693bf
  33:        0x10159859b - rustc::ty::query::<impl rustc::ty::query::config::QueryAccessors for rustc::ty::query::queries::typeck_tables_of>::compute::had7daa30ac2f087b
  34:        0x1015381db - rustc::dep_graph::graph::DepGraph::with_task_impl::h422fc241c771cd19
  35:        0x1015e5f16 - rustc::ty::query::plumbing::<impl rustc::ty::context::TyCtxt>::get_query::ha62632bdf7526a3d
  36:        0x101656962 - rustc_typeck::collect::checked_type_of::h5554750a15f99607
  37:        0x1016560ee - rustc_typeck::collect::type_of::hd12886197ade892b
  38:        0x10150f787 - rustc::ty::query::__query_compute::type_of::hd2086495657a0a1a
  39:        0x10159819b - rustc::ty::query::<impl rustc::ty::query::config::QueryAccessors for rustc::ty::query::queries::type_of>::compute::h1b20e71b2826b327
  40:        0x101542c8b - rustc::dep_graph::graph::DepGraph::with_task_impl::hb48f64bf70c299dd
  41:        0x1015fe09b - rustc::ty::query::plumbing::<impl rustc::ty::context::TyCtxt>::get_query::hc9a002db3551d89f
  42:        0x10166f3f0 - rustc::hir::intravisit::walk_expr::h61b55abdacb61ae7
  43:        0x101638454 - rustc::hir::intravisit::Visitor::visit_fn::h74293bf190427cca
  44:        0x1016720c5 - rustc::hir::intravisit::walk_item::h345dea43b6983efc
  45:        0x101651c78 - <rustc_typeck::collect::CollectItemTypesVisitor as rustc::hir::intravisit::Visitor>::visit_item::h719007420ab8d42f
  46:        0x101510778 - rustc::hir::map::Map::visit_item_likes_in_module::h41a45208ed672977
  47:        0x1016519e4 - rustc_typeck::collect::collect_mod_item_types::hc75ad3d1ce4ddab9
  48:        0x10150eb9a - rustc::ty::query::__query_compute::collect_mod_item_types::h90bdce9721c09aea
  49:        0x10159885b - rustc::ty::query::<impl rustc::ty::query::config::QueryAccessors for rustc::ty::query::queries::collect_mod_item_types>::compute::h1dc8d43facbba52f
  50:        0x10153fa66 - rustc::dep_graph::graph::DepGraph::with_task_impl::h9aef8c33a46ea846
  51:        0x1015fcac2 - rustc::ty::query::plumbing::<impl rustc::ty::context::TyCtxt>::get_query::hc501a1a7cb172f19
  52:        0x10152fa9d - rustc_typeck::check_crate::{{closure}}::{{closure}}::h0075d24b61dd039e
  53:        0x101510ebd - rustc::util::common::time::h2c6325efeb8c917e
  54:        0x1016b58b8 - rustc_typeck::check_crate::h4f0cc6c764300359
  55:        0x10105ce72 - rustc_interface::passes::analysis::h0bbd5fdf3a5c0c97
  56:        0x100f665fe - rustc::ty::query::__query_compute::analysis::h7396128e68e0db7f
  57:        0x100f72b41 - rustc::ty::query::plumbing::<impl rustc::ty::context::TyCtxt>::get_query::h267adeeb0cdf18a0
  58:        0x100f9219c - rustc_interface::passes::BoxedGlobalCtxt::access::{{closure}}::hf2e56cdb05ffd121
  59:        0x1010cf71b - rustc_interface::passes::create_global_ctxt::{{closure}}::h42bdcf19d1a42a9e
  60:        0x100f94352 - rustc_interface::interface::run_compiler_in_existing_thread_pool::he7e390e79c6cfcaf
  61:        0x100fc6284 - std::thread::local::LocalKey<T>::with::h7cc2d0d20721bb64
  62:        0x100fca5c2 - scoped_tls::ScopedKey<T>::set::h1e5f389a2c4779b0
  63:        0x100fe3765 - syntax::with_globals::hccf9e749a1217bc9
  64:        0x100f62d1d - std::sys_common::backtrace::__rust_begin_short_backtrace::h09b527d5f5497ee9
  65:        0x103a40f8f - __rust_maybe_catch_panic
  66:        0x100f87f97 - core::ops::function::FnOnce::call_once{{vtable.shim}}::h4448e4e8ceddd77c
  67:        0x103a1300e - <alloc::boxed::Box<F> as core::ops::function::FnOnce<A>>::call_once::hd1afe4d791e3679f
  68:        0x103a3fd9e - std::sys::unix::thread::Thread::new::thread_start::h785fcf8d81bd0f52
  69:     0x7fff6f0ca2eb - _pthread_body
  70:     0x7fff6f0cd249 - _pthread_start

note: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust/blob/master/CONTRIBUTING.md#bug-reports

note: rustc 1.39.0 (4560ea788 2019-11-04) running on x86_64-apple-darwin

query stack during panic:
thread 'rustc' panicked at 'already borrowed: BorrowMutError', src/libcore/result.rs:1165:5
stack backtrace:
   0:        0x103a2d165 - <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt::h66f38d4abb2e41c1
   1:        0x103a63e10 - core::fmt::write::h4f487e714088986d
   2:        0x103a207bb - std::io::Write::write_fmt::h7bd7b6fe8b47c5fb
   3:        0x103a314aa - std::panicking::default_hook::{{closure}}::heee79636c241547c
   4:        0x103a311b5 - std::panicking::default_hook::hfcbd07059d15441e
   5:        0x100fc4762 - rustc_driver::report_ice::h5ad941de17a16320
   6:        0x103a31ce2 - std::panicking::rust_panic_with_hook::h0c4b67125f55410a
   7:        0x103a3174d - std::panicking::continue_panic_fmt::h0e74ab2b215a1401
   8:        0x103a31649 - rust_begin_unwind
   9:        0x103a5d3bf - core::panicking::panic_fmt::h09741a3213dba543
  10:        0x103a5d5c9 - core::result::unwrap_failed::hf47c31c429b02014
  11:        0x102b1158b - rustc_errors::Handler::force_print_diagnostic::h0bfe3e394dc103fa
  12:        0x102468615 - rustc::ty::query::plumbing::<impl rustc::ty::context::TyCtxt>::try_print_query_stack::ha545746afda6bf84
  13:        0x100fc5392 - rustc_driver::report_ice::h5ad941de17a16320
  14:        0x103a31ce2 - std::panicking::rust_panic_with_hook::h0c4b67125f55410a
  15:        0x102afce51 - std::panicking::begin_panic::h15f5ca0459a0afbf
  16:        0x102b11d37 - rustc_errors::HandlerInner::bug::h3ae5a8494001dbc6
  17:        0x102b10bdd - rustc_errors::Handler::bug::h926b6bc1e5e81222
  18:        0x102424dfb - rustc::util::bug::opt_span_bug_fmt::{{closure}}::hc9e1ba37629e505a
  19:        0x10241f306 - rustc::ty::context::tls::with_opt::{{closure}}::h0f2d2c67a51c6ea3
  20:        0x10241f294 - rustc::ty::context::tls::with_context_opt::hffb989149536908e
  21:        0x10241f2c2 - rustc::ty::context::tls::with_opt::h8dbbf9943655c843
  22:        0x102424d18 - rustc::util::bug::opt_span_bug_fmt::haea6cb90fd22dc2a
  23:        0x102424c6b - rustc::util::bug::bug_fmt::hd6a2040ffa44f924
  24:        0x102616caf - rustc::ty::context::TypeckTables::node_type::{{closure}}::h7013b08e9165475c
  25:        0x102617263 - rustc::ty::context::TypeckTables::expr_ty::h97963f0b78948fbd
  26:        0x10165102e - <rustc_typeck::check::generator_interior::InteriorVisitor as rustc::hir::intravisit::Visitor>::visit_expr::h3b58dc6fbd3e7928
  27:        0x10166ccbf - rustc::hir::intravisit::walk_expr::h05f1735da88485b2
  28:        0x101650f7c - <rustc_typeck::check::generator_interior::InteriorVisitor as rustc::hir::intravisit::Visitor>::visit_expr::h3b58dc6fbd3e7928
  29:        0x101650f7c - <rustc_typeck::check::generator_interior::InteriorVisitor as rustc::hir::intravisit::Visitor>::visit_expr::h3b58dc6fbd3e7928
  30:        0x101650f7c - <rustc_typeck::check::generator_interior::InteriorVisitor as rustc::hir::intravisit::Visitor>::visit_expr::h3b58dc6fbd3e7928
  31:        0x101650f7c - <rustc_typeck::check::generator_interior::InteriorVisitor as rustc::hir::intravisit::Visitor>::visit_expr::h3b58dc6fbd3e7928
  32:        0x10165061d - rustc_typeck::check::generator_interior::resolve_interior::h94e9b683be16f5d7
  33:        0x1014ce8b1 - rustc_typeck::check::FnCtxt::resolve_generator_interiors::h3f88be1dbc9c6168
  34:        0x10161c575 - rustc::ty::context::GlobalCtxt::enter_local::h5020ef7c3f2707ac
  35:        0x1014c0a5c - rustc_typeck::check::typeck_tables_of::h1d829091428d57d9
  36:        0x10150dcda - rustc::ty::query::__query_compute::typeck_tables_of::h45dc00b6674693bf
  37:        0x10159859b - rustc::ty::query::<impl rustc::ty::query::config::QueryAccessors for rustc::ty::query::queries::typeck_tables_of>::compute::had7daa30ac2f087b
  38:        0x1015381db - rustc::dep_graph::graph::DepGraph::with_task_impl::h422fc241c771cd19
  39:        0x1015e5f16 - rustc::ty::query::plumbing::<impl rustc::ty::context::TyCtxt>::get_query::ha62632bdf7526a3d
  40:        0x10150dcda - rustc::ty::query::__query_compute::typeck_tables_of::h45dc00b6674693bf
  41:        0x10159859b - rustc::ty::query::<impl rustc::ty::query::config::QueryAccessors for rustc::ty::query::queries::typeck_tables_of>::compute::had7daa30ac2f087b
  42:        0x1015381db - rustc::dep_graph::graph::DepGraph::with_task_impl::h422fc241c771cd19
  43:        0x1015e5f16 - rustc::ty::query::plumbing::<impl rustc::ty::context::TyCtxt>::get_query::ha62632bdf7526a3d
  44:        0x101656962 - rustc_typeck::collect::checked_type_of::h5554750a15f99607
  45:        0x1016560ee - rustc_typeck::collect::type_of::hd12886197ade892b
  46:        0x10150f787 - rustc::ty::query::__query_compute::type_of::hd2086495657a0a1a
  47:        0x10159819b - rustc::ty::query::<impl rustc::ty::query::config::QueryAccessors for rustc::ty::query::queries::type_of>::compute::h1b20e71b2826b327
  48:        0x101542c8b - rustc::dep_graph::graph::DepGraph::with_task_impl::hb48f64bf70c299dd
  49:        0x1015fe09b - rustc::ty::query::plumbing::<impl rustc::ty::context::TyCtxt>::get_query::hc9a002db3551d89f
  50:        0x10166f3f0 - rustc::hir::intravisit::walk_expr::h61b55abdacb61ae7
  51:        0x101638454 - rustc::hir::intravisit::Visitor::visit_fn::h74293bf190427cca
  52:        0x1016720c5 - rustc::hir::intravisit::walk_item::h345dea43b6983efc
  53:        0x101651c78 - <rustc_typeck::collect::CollectItemTypesVisitor as rustc::hir::intravisit::Visitor>::visit_item::h719007420ab8d42f
  54:        0x101510778 - rustc::hir::map::Map::visit_item_likes_in_module::h41a45208ed672977
  55:        0x1016519e4 - rustc_typeck::collect::collect_mod_item_types::hc75ad3d1ce4ddab9
  56:        0x10150eb9a - rustc::ty::query::__query_compute::collect_mod_item_types::h90bdce9721c09aea
  57:        0x10159885b - rustc::ty::query::<impl rustc::ty::query::config::QueryAccessors for rustc::ty::query::queries::collect_mod_item_types>::compute::h1dc8d43facbba52f
  58:        0x10153fa66 - rustc::dep_graph::graph::DepGraph::with_task_impl::h9aef8c33a46ea846
  59:        0x1015fcac2 - rustc::ty::query::plumbing::<impl rustc::ty::context::TyCtxt>::get_query::hc501a1a7cb172f19
  60:        0x10152fa9d - rustc_typeck::check_crate::{{closure}}::{{closure}}::h0075d24b61dd039e
  61:        0x101510ebd - rustc::util::common::time::h2c6325efeb8c917e
  62:        0x1016b58b8 - rustc_typeck::check_crate::h4f0cc6c764300359
  63:        0x10105ce72 - rustc_interface::passes::analysis::h0bbd5fdf3a5c0c97
  64:        0x100f665fe - rustc::ty::query::__query_compute::analysis::h7396128e68e0db7f
  65:        0x100f72b41 - rustc::ty::query::plumbing::<impl rustc::ty::context::TyCtxt>::get_query::h267adeeb0cdf18a0
  66:        0x100f9219c - rustc_interface::passes::BoxedGlobalCtxt::access::{{closure}}::hf2e56cdb05ffd121
  67:        0x1010cf71b - rustc_interface::passes::create_global_ctxt::{{closure}}::h42bdcf19d1a42a9e
  68:        0x100f94352 - rustc_interface::interface::run_compiler_in_existing_thread_pool::he7e390e79c6cfcaf
  69:        0x100fc6284 - std::thread::local::LocalKey<T>::with::h7cc2d0d20721bb64
  70:        0x100fca5c2 - scoped_tls::ScopedKey<T>::set::h1e5f389a2c4779b0
  71:        0x100fe3765 - syntax::with_globals::hccf9e749a1217bc9
  72:        0x100f62d1d - std::sys_common::backtrace::__rust_begin_short_backtrace::h09b527d5f5497ee9
  73:        0x103a40f8f - __rust_maybe_catch_panic
  74:        0x100f87f97 - core::ops::function::FnOnce::call_once{{vtable.shim}}::h4448e4e8ceddd77c
  75:        0x103a1300e - <alloc::boxed::Box<F> as core::ops::function::FnOnce<A>>::call_once::hd1afe4d791e3679f
  76:        0x103a3fd9e - std::sys::unix::thread::Thread::new::thread_start::h785fcf8d81bd0f52
  77:     0x7fff6f0ca2eb - _pthread_body
  78:     0x7fff6f0cd249 - _pthread_start

error: internal compiler error: unexpected panic

note: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust/blob/master/CONTRIBUTING.md#bug-reports

note: rustc 1.39.0 (4560ea788 2019-11-04) running on x86_64-apple-darwin

query stack during panic:
thread panicked while processing panic. aborting.
[1]    63445 illegal hardware instruction  RUSTC_LOG=ty::context=debug RUST_BACKTRACE=full rustc foo.rs
```

**Comment by jonas-schievink** [comment]
Please check on the newest nightly, I think this is already fixed there.

**Comment by nate-onesignal** [comment]
Yes, it looks like this works as expected in the newest nightly
